torch/csrc/utils/tensor_flatten.h

```diff
@@ -9,10 +9,22 @@
 namespace torch { namespace utils {
 
 inline at::Tensor flatten_dense_tensors(at::TensorList tensors) {
-  static auto flatten = [](const at::Tensor &t) { return t.contiguous().view({-1}); };
-  if (tensors.size() == 1)
-    return flatten(tensors[0]);
-  return at::cat(fmap(tensors, flatten));
+  if (tensors.size() == 1) {
+    return tensors[0].reshape({-1});
+  } else {
+    int64_t total_numel = 0;
+    for (const auto & tensor : tensors) {
+      total_numel += tensor.numel();
+    }
+    auto flat = tensors[0].type().tensor({total_numel});
+    int64_t offset = 0;
+    for (const auto & tensor : tensors) {
+      auto numel = tensor.numel();
+      flat.narrow(0, offset, numel).view_as(tensor).copy_(tensor);
```

Comment:
Isn't this less efficient on the GPU than what we had before with the `at::cat`?

Reply:
It is. `at::cat` has specialized kernels for this thing, that can be faster than triggering a separate copy.